lib/editor/actions/map/stopStrategies.js

```diff
@@ -286,48 +283,64 @@ export function addStopToPattern (pattern: Pattern, stop: GtfsStop, index?: ?num
+        // behavior in this code block essentially replaces any
+        // non-stop control points in the "from" segment with the new pattern stop and re-routes the
+        // shape between it and the new stop.
+        // Non-stop control points are preserved in the new "to" segment.
+        //     ↙️ "from" segment      ↙️ "to" segment
+        //  0 ——x——— 0 ——————————x———x——— 0
+        //             ^— 0 —^ <--- new inserted stop
+
         // Add control point in order to copy of current list.
-        const controlPoint = newControlPoint(distanceInMeters, insertPoint, {
-          stopId: stop.stop_id,
-          pointType: POINT_TYPE.STOP // 2
-        })
         const stopControlPoints = controlPoints
           // TODO: refactor into shared function (see pattern-debug-lines.js)
           .map((cp, index) => ({...cp, cpIndex: index}))
           .filter(cp => cp.pointType === POINT_TYPE.STOP)
-        const {cpIndex: nextStopIndex} = stopControlPoints[index]
-        // Iterate over control points to find previous and next stop control
-        // points.
-        let spliceIndex = 1 // Init to 1 to avoid crashing in a case where we can't project the stop onto the existing pattern.
-        while (controlPoints[spliceIndex].distance < distanceInMeters && spliceIndex < nextStopIndex) {
-          spliceIndex++
-        }
         // Perform splice operation on cloned control points to remove any
         // control points between the previous and next stop control points and
         // insert new stop control point in their stead.
         const clonedControlPoints = clone(controlPoints)
         const clonedPatternSegments = clone(patternSegments)
-        // Replace n segments with 2 blank "placeholder" segments to be replaced
+
+        // Insert the new stop after the previous stopControlPoint.
+        // We assume that all control points hold for the "to" semgent (see above).
+        const previousStopControlPoint = stopControlPoints[index - 1]
+        const spliceIndex = previousStopControlPoint.cpIndex + 1
+        const nextControlPoint = clonedControlPoints[spliceIndex]
+
+        // Previously, at this point we had created our control point by projecting the stop onto the existing shape.
+        // However, this created problems for insertions where the stop is geographically before the previous stop
+        // but is being inserted after it in the stop sequence. We were projecting previously bc to create a control point
+        // we need to know the distance along the shape, but to avoid the above issue we need to make a sub-request to graphhopper
+        // to get the segment between the two points and from that, the distance.
+        const insertPoint = stopToGeoJSONPoint(stop)
+        const newFromSegmentCoords = [
+          previousStopControlPoint.point.geometry.coordinates,
+          [stop.stop_lon, stop.stop_lat]
+        ]
+        const newFromSegment = await getSegment(newFromSegmentCoords, true)
```

Comment:
This needs some null checks

Reply:
how bout just tossing it in the try block since it's the same call as recalculateShapes effectively?
https://github.com/ibi-group/datatools-ui/pull/918/commits/5f9c7d16521a85c0832fc4349c509460e16ab0fc